lib/core.js

```diff
@@ -407,7 +407,7 @@ declare class Promise<R> {
 
     then<U>(
       onFulfill?: (value: R) => Promise<U> | U,
-      onReject?: (error: any) => Promise<U> | U
+      onReject?: (error: any) => ?Promise<U> | U
```

Comment:
Ok, I'm starting to believe that my proposed change is incorrect: the `onReject` callback in `then` should remain as is, whereas the callback in `catch` must be fixed (converted to non-nullable).
The idea is that if a promise fails, and there is a catch block that returns nothing, then the after the error is caught the promise is assumed successful, returning the result of the last onReject callback.

Here's a test case:
```
/* @flow */

function promiseMeMore(promise: Promise<number>): Promise<number> {
  return promise.then((x) => 2*x,
                      () => console.log("You broke your promise!"));
}
function boast(x: number) {
  console.assert(typeof(x) === "number", "Oops, got "+typeof(x)+" instead of a number");
  console.log("You promised me: " + (1*x));
  return x;
}

var promise1: Promise<number> = Promise.resolve(100).then(boast);
var promise2: Promise<number> = promiseMeMore(promise1).then(boast);
var promise3: Promise<number> = Promise.reject().then(boast);
var promise4: Promise<number> = promiseMeMore(promise3).then(boast);
```
In this test `promise4` is invalid, trying to execute it will hit an assertion in `boast()`. For whatever reason I can't make Flow to catch this problem, with either old or new definitions...